Mismatched type parameter nullability

Upgrading from `0.12.1` to `0.12.2` and I am confused by these new warnings.

```console
> Task :caffeine:compileJava
/Users/ben/projects/caffeine/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java:3180: warning: [NullAway] Cannot pass parameter of type Function<@org.jspecify.annotations.Nullable V, @org.jspecify.annotations.Nullable V>, as formal parameter has type Function<V, V>, which has mismatched type parameter nullability
            .map(node -> nodeToCacheEntry(node, transformer))
                                                ^
    (see http://t.uber.com/nullaway )
  Did you mean '@SuppressWarnings("NullAway") <T> T snapshot(Iterable<Node<K, V>> iterable, Function<@Nullable V, @Nullable V> transformer,'?
```
```console
/Users/ben/projects/caffeine/caffeine/src/main/java/com/github/benmanes/caffeine/cache/BoundedLocalCache.java:4014: warning: [NullAway] Cannot pass parameter of type Function<@org.jspecify.annotations.Nullable V, @org.jspecify.annotations.Nullable V>, as formal parameter has type Function<V, V>, which has mismatched type parameter nullability
      return (node == null) ? null : cache.nodeToCacheEntry(node, transformer);
                                                                  ^
    (see http://t.uber.com/nullaway )
  Did you mean '@SuppressWarnings("NullAway") @Override public @Nullable CacheEntry<K, V> getEntryIfPresentQuietly(K key) {'?
```

The `snapshot` method and `nodeToCacheEntry` method both use the same types for the `transformer` parameter.
```java
  <T> T snapshot(Iterable<Node<K, V>> iterable, Function<@Nullable V, @Nullable V> transformer,
      Function<Stream<CacheEntry<K, V>>, T> mappingFunction)

  @Nullable CacheEntry<K, V> nodeToCacheEntry(
      Node<K, V> node, Function<@Nullable V, @Nullable V> transformer)
```

The other warning is similar. It appears to be a regression.

Note that you might have to use `--rerun`, `--rerun-tasks` or `--no-build-cache` to force recompilations.

--- a/nullaway/src/main/java/com/uber/nullaway/dataflow/CoreNullnessStoreInitializer.java
+++ b/nullaway/src/main/java/com/uber/nullaway/dataflow/CoreNullnessStoreInitializer.java
@@ -1,5 +1,6 @@
 package com.uber.nullaway.dataflow;
 
+import static com.uber.nullaway.NullabilityUtil.castToNonNull;
 import static com.uber.nullaway.Nullness.NONNULL;
 import static com.uber.nullaway.Nullness.NULLABLE;
 
@@ -105,7 +106,8 @@ class CoreNullnessStoreInitializer extends NullnessStoreInitializer {
     // This obtains the types of the functional interface method parameters with preserved
     // annotations in case of generic type arguments.  Only used in JSpecify mode.
     List<Type> overridenMethodParamTypeList =
-        TypeSubstitutionUtils.memberType(types, ASTHelpers.getType(code), fiMethodSymbol)
+        TypeSubstitutionUtils.memberType(
+                types, castToNonNull(ASTHelpers.getType(code)), fiMethodSymbol)
             .getParameterTypes();
     // If fiArgumentPositionNullness[i] == null, parameter position i is unannotated
     Nullness[] fiArgumentPositionNullness = new Nullness[fiMethodParameters.size()];

--- a/nullaway/src/main/java/com/uber/nullaway/dataflow/CoreNullnessStoreInitializer.java
+++ b/nullaway/src/main/java/com/uber/nullaway/dataflow/CoreNullnessStoreInitializer.java
@@ -16,6 +16,7 @@ import com.uber.nullaway.CodeAnnotationInfo;
 import com.uber.nullaway.Config;
 import com.uber.nullaway.NullabilityUtil;
 import com.uber.nullaway.Nullness;
+import com.uber.nullaway.generics.TypeSubstitutionUtils;
 import com.uber.nullaway.handlers.Handler;
 import java.util.List;
 import java.util.Objects;
@@ -104,7 +105,8 @@ class CoreNullnessStoreInitializer extends NullnessStoreInitializer {
     // This obtains the types of the functional interface method parameters with preserved
     // annotations in case of generic type arguments.  Only used in JSpecify mode.
     List<Type> overridenMethodParamTypeList =
-        types.memberType(ASTHelpers.getType(code), fiMethodSymbol).getParameterTypes();
+        TypeSubstitutionUtils.memberType(types, ASTHelpers.getType(code), fiMethodSymbol)
+            .getParameterTypes();
     // If fiArgumentPositionNullness[i] == null, parameter position i is unannotated
     Nullness[] fiArgumentPositionNullness = new Nullness[fiMethodParameters.size()];
     boolean isFIAnnotated =

--- a/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
+++ b/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
@@ -26,6 +26,7 @@ import com.sun.tools.javac.code.Attribute;
 import com.sun.tools.javac.code.Symbol;
 import com.sun.tools.javac.code.TargetType;
 import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Types;
 import com.sun.tools.javac.tree.JCTree;
 import com.uber.nullaway.CodeAnnotationInfo;
 import com.uber.nullaway.Config;
@@ -624,7 +625,8 @@ public final class GenericsChecks {
         enclosingType = methodSymbol.owner.type;
       }
       if (enclosingType != null) {
-        invokedMethodType = state.getTypes().memberType(enclosingType, methodSymbol);
+        invokedMethodType =
+            TypeSubstitutionUtils.memberType(state.getTypes(), enclosingType, methodSymbol);
       }
     }
     // substitute type arguments for generic methods
@@ -694,7 +696,8 @@ public final class GenericsChecks {
     // Obtain type parameters for the overridden method within the context of the overriding
     // method's class
     Type methodWithTypeParams =
-        state.getTypes().memberType(overridingMethod.owner.type, overriddenMethod);
+        TypeSubstitutionUtils.memberType(
+            state.getTypes(), overridingMethod.owner.type, overriddenMethod);
 
     checkTypeParameterNullnessForOverridingMethodReturnType(
         tree, methodWithTypeParams, analysis, state);
@@ -772,7 +775,8 @@ public final class GenericsChecks {
       // annotation should have been handled by the caller)
       return Nullness.NONNULL;
     }
-    Type overriddenMethodType = state.getTypes().memberType(enclosingType, method);
+    Type overriddenMethodType =
+        TypeSubstitutionUtils.memberType(state.getTypes(), enclosingType, method);
     verify(
         overriddenMethodType instanceof ExecutableType,
         "expected ExecutableType but instead got %s",
@@ -873,7 +877,9 @@ public final class GenericsChecks {
 
     Type.ForAll forAllType = (Type.ForAll) methodSymbol.type;
     Type.MethodType underlyingMethodType = (Type.MethodType) forAllType.qtype;
-    return state.getTypes().subst(underlyingMethodType, forAllType.tvars, explicitTypeArgs);
+    Types types = state.getTypes();
+    return TypeSubstitutionUtils.subst(
+        types, underlyingMethodType, forAllType.tvars, explicitTypeArgs);
   }
 
   /**
@@ -1007,7 +1013,7 @@ public final class GenericsChecks {
       // @Nullable annotation is handled elsewhere)
       return Nullness.NONNULL;
     }
-    Type methodType = state.getTypes().memberType(enclosingType, method);
+    Type methodType = TypeSubstitutionUtils.memberType(state.getTypes(), enclosingType, method);
     Type paramType = methodType.getParameterTypes().get(parameterIndex);
     return getTypeNullness(paramType, config);
   }

--- a/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
+++ b/nullaway/src/main/java/com/uber/nullaway/generics/GenericsChecks.java
@@ -26,7 +26,6 @@ import com.sun.tools.javac.code.Attribute;
 import com.sun.tools.javac.code.Symbol;
 import com.sun.tools.javac.code.TargetType;
 import com.sun.tools.javac.code.Type;
-import com.sun.tools.javac.code.Types;
 import com.sun.tools.javac.tree.JCTree;
 import com.uber.nullaway.CodeAnnotationInfo;
 import com.uber.nullaway.Config;
@@ -877,9 +876,8 @@ public final class GenericsChecks {
 
     Type.ForAll forAllType = (Type.ForAll) methodSymbol.type;
     Type.MethodType underlyingMethodType = (Type.MethodType) forAllType.qtype;
-    Types types = state.getTypes();
     return TypeSubstitutionUtils.subst(
-        types, underlyingMethodType, forAllType.tvars, explicitTypeArgs);
+        state.getTypes(), underlyingMethodType, forAllType.tvars, explicitTypeArgs);
   }
 
   /**

--- a/nullaway/src/main/java/com/uber/nullaway/generics/TypeSubstitutionUtils.java
+++ b/nullaway/src/main/java/com/uber/nullaway/generics/TypeSubstitutionUtils.java
@@ -1,0 +1,35 @@
+package com.uber.nullaway.generics;
+
+import com.sun.tools.javac.code.Symbol;
+import com.sun.tools.javac.code.Type;
+import com.sun.tools.javac.code.Types;
+import com.sun.tools.javac.util.List;
+
+/** Utility method related to substituting type arguments for type variables. */
+public class TypeSubstitutionUtils {
+
+  /**
+   * Returns the type of {@code sym} as a member of {@code t}.
+   *
+   * @param types the {@link Types} instance
+   * @param t the enclosing type
+   * @param sym the symbol
+   * @return the type of {@code sym} as a member of {@code t}
+   */
+  public static Type memberType(Types types, Type t, Symbol sym) {
+    return types.memberType(t, sym);
+  }
+
+  /**
+   * Substitutes the types in {@code to} for the types in {@code from} in {@code t}.
+   *
+   * @param types the {@link Types} instance
+   * @param t the type to which to perform the substitution
+   * @param from the types that will be substituted out
+   * @param to the types that will be substituted in
+   * @return the type resulting from the substitution
+   */
+  public static Type subst(Types types, Type t, List<Type> from, List<Type> to) {
+    return types.subst(t, from, to);
+  }
+}

